File: sections/21_graphs_i_mixed_contest/problems/a_existing_network/solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct DSU {
    vector<int> parent, sz;
    DSU(int n) : parent(n), sz(n, 1) {
        iota(parent.begin(), parent.end(), 0);
    }
    int find(int x) {
        while (parent[x] != x) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }
    bool unite(int a, int b) {
        a = find(a);
        b = find(b);
        if (a == b) return false;
        if (sz[a] < sz[b]) swap(a, b);
        parent[b] = a;
        sz[a] += sz[b];
        return true;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, f, p;
    if (!(cin >> n >> f >> p)) return 0;
    DSU dsu(n);
    int components = n;
    for (int i = 0; i < f; i++) {
        int u, v;
        cin >> u >> v;
        if (dsu.unite(u - 1, v - 1)) components--;
    }
    vector<tuple<long long,int,int>> edges;
    for (int i = 0; i < p; i++) {
        int u, v;
        long long w;
        cin >> u >> v >> w;
        edges.push_back({w, u - 1, v - 1});
    }
    sort(edges.begin(), edges.end());
    long long total = 0;
    for (auto [w, u, v] : edges) {
        if (components == 1) break;
        if (dsu.unite(u, v)) {
            total += w;
            components--;
        }
    }
    if (components == 1) cout << total << '\n';
    else cout << "IMPOSSIBLE\n";
    return 0;
}

File: sections/21_graphs_i_mixed_contest/problems/a_existing_network/solution.py
import sys


class DSU:
    def __init__(self, n: int) -> None:
        self.parent = list(range(n))
        self.size = [1] * n

    def find(self, x: int) -> int:
        while self.parent[x] != x:
            self.parent[x] = self.parent[self.parent[x]]
            x = self.parent[x]
        return x

    def union(self, a: int, b: int) -> bool:
        a = self.find(a)
        b = self.find(b)
        if a == b:
            return False
        if self.size[a] < self.size[b]:
            a, b = b, a
        self.parent[b] = a
        self.size[a] += self.size[b]
        return True


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    it = iter(data)
    n = next(it)
    f = next(it)
    p = next(it)
    dsu = DSU(n)
    components = n
    for _ in range(f):
        u = next(it) - 1
        v = next(it) - 1
        if dsu.union(u, v):
            components -= 1
    edges = []
    for _ in range(p):
        u = next(it) - 1
        v = next(it) - 1
        w = next(it)
        edges.append((w, u, v))
    edges.sort()
    total = 0
    for w, u, v in edges:
        if components == 1:
            break
        if dsu.union(u, v):
            total += w
            components -= 1
    print(total if components == 1 else "IMPOSSIBLE")


if __name__ == "__main__":
    main()

